File: 0188. Best Time to Buy and Sell Stock IV.cpp
class Solution
{
public:
    int maxProfit(int k, vector<int>& prices)
    {
        if (k == 0)
        {
            return 0;
        }
        
        vector<int> minCost(k + 1, INT_MAX);
        vector<int> maxProfit(k + 1);
        
        for (int price : prices)
        {
            for (int i = 1; i <= k; ++i)
            {
                minCost[i] = min(minCost[i], price - maxProfit[i - 1]);
                maxProfit[i] = max(maxProfit[i], price - minCost[i]);
            }
        }
        
        return maxProfit[k];
    }
};

File: 0188. Best Time to Buy and Sell Stock IV.cs
public class Solution
{
    public int MaxProfit(int k, int[] prices)
    {
        var minPrice = Enumerable.Range(0, k).Select(x => int.MaxValue).ToArray();
        var maxProfit = Enumerable.Range(0, k).Select(x => 0).ToArray();

        foreach (var price in prices)
        {
            for (int i = 0; i < k; ++i)
            {
                var previousProfit = i > 0 ? maxProfit[i - 1] : 0;
                minPrice[i] = Math.Min(minPrice[i], price - previousProfit);
                maxProfit[i] = Math.Max(maxProfit[i], price - minPrice[i]);
            }
        }

        return maxProfit[^1];
    }
}